### PR TITLE
feat: add filtering + search to ecosystem page

### LINKED ISSUE #177: [Ecosystem page] Implement dynamic filtering and search
Note: [static page](https://github.com/safe-global/safe-homepage/issues/176) needs to be created first.

## What is the feature about

Allowing search/filtering of project list.

## The list of requirements

1. Port search/filtering logic/packages from `web-core` (Fuse, animation package, etc.) to landing page.

- [x] Implement search
- [x] Implement filter
- [x] Implement pagination

## Links

- [Epic](https://github.com/safe-global/safe-homepage/issues/161).

### PR DESCRIPTION
**Pull request by iamacook**
Resolves #177 by adding the ability to filter and search the projects on the Ecosystem page.

image

Screenshot 2023-05-04 at 19 43 38

Screenshot 2023-05-04 at 19 43 47

image

### CODE REVIEW

**Comment by katspaugh** [comment]
* Checkbox labels should be clickable as well as checkboxes themselves (use `<label>`)
* Evaluate if the filtering should be finding overlapping sets instead of the sum of tags. E.g. if I selected DEX and NFTs, should it find only the apps that are BOTH DEXes and NFT apps, or DEX apps and NFT apps together? I would expect the former.
* Next time, please break down the work into small components. E.g. the tag sidebar separately, search separately, then one integration PR.

**Comment by iamacook** [comment]
> Checkbox labels should be clickable as well as checkboxes themselves (use `<label>`)

It's now possible to click on the labels to (un-)check the checkboxes.

> Evaluate if the filtering should be finding overlapping sets instead of the sum of tags. E.g. if I selected DEX and NFTs, should it find only the apps that are BOTH DEXes and NFT apps, or DEX apps and NFT apps together? I would expect the former.

I've altered it so that the results must now include all selected filters.

**Review by usame-algan — APPROVED** [review]
Great job, very smooth UX! 👍

File: src/components/Ecosystem/Projects/useProjectsSearch.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+          },
+          {
+            name: 'primary_category',
+            weight: 0.5,
+          },
+          {
+            name: 'secondary_categories',
+            weight: 0.2,
+          },
+        ],
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/Ecosystem/Projects/useProjectsSearch.ts:27]: The `weight` is how [high up they rank](https://fusejs.io/examples.html#weighted-search). I'll remove the category/integration search in place of description though.

File: src/components/Ecosystem/Projects/index.tsx [hunk 2]
```diff
@@ -40,107 +51,273 @@ type EcosystemProject = {
...
+
+          <Grid item xs={12} md={9} className={css.chipContainer}>
+            {selectedCategories.map((category) => (
+              <Chip
+                key={category}
+                className={css.chip}
+                label={category}
+                onDelete={() => onSelectCategory(category, false)}
+                deleteIcon={<CrossIcon />}
+              />
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/components/Ecosystem/Projects/index.tsx:238]: They do - tab to a `Chip` and [press Delete](https://github.com/mui/material-ui/issues/10598).

**Pull request merged**